app/custom_auth.py

```diff
@@ -0,0 +1,106 @@
+from django.contrib import auth
+from django.conf import settings
+from django.contrib.auth.models import User, Group
+from django.core.exceptions import PermissionDenied
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+class CustomSSOMiddleware:
+    """
+    A custom middleware that allows SSO via HTTP headers passed in the request
+    
+    Example 
+    -------
+    HTTP_X_AUTH_PROXY_USER : User 
+    HTTP_X_AUTH_PROXY_GROUPS: Admin,Editor
+
+    Summary
+    -------
+    (1) If HTTP header HTTP_X_AUTH_PROXY_USER with the user name is not passed no further processing is done by the middleware.
+    (2) If HTTP header with the user name is passed then:
+        (2.1) Authenciate the user if already exist in the database. 
+        (2.2) Create user if does not exist in the database and then autenicate the user. 
+    (3) If IS_SUPER flag is set to True in the settings.py then ingnore HTTP_X_AUTH_PROXY_GROUPS and grant the user Admin level privileges
+    (4) If HTTP_X_AUTH_PROXY_GROUPS header is present and contains Admin group then grant the user Admin level privileges. 
+    (5) If HTTP_X_AUTH_PROXY_GROUPS header is absent or does not contain the Admin group then revoke user's Admin level privleges.  
+    (6) User is now signed in and session cookies are set.  
+    ...
+    """
+    def __init__(self, get_response):
+        self.get_response = get_response
+     
+    def __call__(self, request):  
+        try:
+            username = request.META[settings.HEADER_AUTH_PROXY_USER]
+            
+            if not User.objects.filter(username=username).exists() and settings.CREATE_USER:
+                create_user(username)
+            
+            user = auth.authenticate(remote_user=username)
+            process_user_groups(user,request.META)
+            
+            if user:
+                request.user = user
+                auth.login(request, user)
+
+        except KeyError:
+            logger.info(f"LOG: User name header {settings.HEADER_AUTH_PROXY_USER} not found. Continue without user authenication." )
+         
+        response = self.get_response(request)
+        return response
+
+def process_user_groups(user,headers):
```

Comment:
Not sure I understand what's going on here. Currently doccano doesn't have any groups beyond super-user. To keep the implementation simple, I'd recommend the following:

1) Parse the header which contains the group entitlements (e.g. split on `,`).
2) If any of the group entitlements contains the `settings.ADMIN_GROUP_NAME`, set the user's superuser flag to true. Otherwise set it to false.

Reply:
```superuser`` is a [special property ](https://docs.djangoproject.com/en/2.2/ref/django-admin/#django-admin-createsuperuser) of the user object inside Django that gives the user all permissions available in the system. It is akin to root in Linux.  I won't recommend using it broadly as it encourages poor security practices.  

If the intention is to keep permission design simple then that is already taken care of by the code.  Basically, you pass Admin as a group in the header, the user will be assigned  ```superuser`` permissions along with the assignment to the Admin group. In the future when more groups are added it would be much easier to scale this design rather than bucketing every admin user as  a ```superuser``

Reply:
There is no admin role in doccano currently (all the checks go against superuser) so we should work with this approach that's already in the project. Once more roles get rolled out in doccano, this can always be refactored trivially.

Reply:
The code already takes care of what doccano currently supports. Can you point me to the code which will break the existing doccano authorization checks/logic? 

Also, keep in mind this is custom middleware so they can always turn it off as/if needed.

Reply:
There's lots of usage of `is_superuser` in the current doccano code base, most notably in the [User serializer](https://github.com/chakki-works/doccano/blob/50c7078c16e69575d069f110b7a39925b5f732c3/app/api/serializers.py#L16) which communicates the user's roles to the frontend. As such, for now we should only care about `is_superuser` and no other permissions. Once doccano implements a more thorough permissions system, we can refactor the middleware to support that, but for now I'd state that we should keep the system as simple as possible to support only the functionality that's currently being used in doccano (otherwise we essentially introduce a [lava layer](http://mikehadlow.blogspot.com/2014/12/the-lava-layer-anti-pattern.html) of inconsistency in terms of how permissions are handled).

Reply:
Concrete example, these are the only two scenarios we need to support right now:

```
X-User: clewolff
X-Groups: Admin
 
--> is_super
 
X-User: scicoria
X-Groups: ''
 
--> not is_super
```